conn.go

```diff
@@ -164,26 +157,42 @@ func (c *Conn) close() error {
 	return err
 }
 
-func (c *Conn) timeoutLoop() {
-	defer close(c.timeoutLoopDone)
+func (c *Conn) setupWriteTimeout(ctx context.Context) {
+	hammerTime := context.AfterFunc(ctx, func() {
+		c.close()
```

Comment:
I wonder if it'd be possible for this function to also remove itself from `c.writeTimeoutCloser`? Perhaps it can put a `nil` in it / or swap it with a `nil` and run what was in it if it's not `nil`. That way we'll let GC collect the function once its ran. Same below.

Well, it just needs to call `clearWriteTimeout()`.